### PR TITLE
Support string castings for ORC reading [databricks]

### LINKED ISSUE #6394: [FEA] Support string castings for ORC reading
A sub-issue of https://github.com/NVIDIA/spark-rapids/issues/6149 .

While reading ORC files, implement casting `string` to `{bool, int8/16/32/64, float32, double, timestamp, date}`.

### PR DESCRIPTION
**Pull request by sinkinben**
Close #6394 (which is sub-issue of #6149 ).

While reading an ORC file in spark-rapids, we may need to cast `string` to `{bool, int8/16/32/64, float, double, timestamp, date}`.

Now, I have implemented `string -> bool, int types, float`.

### CODE REVIEW

**Comment by sinkinben** [comment]
Here are some corner case that we should pay attention to.
```scala
val data = Seq(
    ("", "123"),    // valid
    ("", "123.0"),  // valid
    ("", "123."),   // valid, 123.0
    ("", "-0123."),  // valid, -123.0
    ("", ".123"),   // valid, 0.123
    ("", "-.123"),  // valid, -0.123
    ("", "+0.0"),   // valid
    ("", "-0.0"),   // valid
    ("", "000123.00"), // valid
    ("", "  10.0"),    // valid
    ("", "    "),      // invalid, replace it with null
    ("", "+1.2e3"),    // valid
    ("", "-3.14e-10   "),  // valid
    ("", "-3.14E-10"),     // valid
    ("", "+03.14e0010"),   // valid
    ("", "-.123e001"),     // valid -1.23
    ("", "00314e0012"),    // valid
    ("", "inf"),           // invalid
    ("", "  Infinity  "),  // valid
    ("", "+Infinity"),     // valid
    ("", "-Infinity"),     // valid
    ("", "NaN"),           // valid
    ("", "-NaN"),          // valid, still NaN
    ("", "nan"),           // invalid
    ("", "9.99e99999"),    // overflow float_max, convert to Infinity
    ("", "0.00001e-9999"), // overflow float_min, convert to 0.0
    ("", "true"),          // invalid, null
    ("", "abc"),           // invalid, null
    ("", "."),             // invalid, null
    ("", "1.2."),          // invalid, null
    ("", "1.2.3")
)
val cols = Seq("str", "str_float")
val df=spark.createDataFrame(data).toDF(cols:_*).repartition(1)
df.printSchema()
df.show(100)
df.write.mode("overwrite").orc("/tmp/orc/data.orc")
```

And we cast `str_float` to `float`, the result of CPU spark is:
```shell
scala> var df = spark.read.schema("str_float float").orc("/tmp/orc/data.orc").show(100)
+-------------+
|    str_float|
+-------------+
|        123.0|
|        123.0|
|        123.0|
|       -123.0|
|        0.123|
|       -0.123|
|          0.0|
|         -0.0|
|        123.0|
|         10.0|
|         null|
|       1200.0|
|    -3.14E-10|
|    -3.14E-10|
|3.13999995E10|
|        -1.23|
|3.13999992E14|
|         null|
|     Infinity|
|     Infinity|
|    -Infinity|
|          NaN|
|          NaN|
|         null|
|     Infinity|
|          0.0|
|         null|
|         null|
|         null|
|         null|
|         null|
+-------------+
```

File: integration_tests/src/main/python/orc_cast_string_test.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ def test_casting_string_to_integers(spark_tmp_path, to_type):
     orc_path = spark_tmp_path + '/orc_cast_string_to_int'
     normal_cases = []
     length = 2048
-    while len(normal_cases) < length:
+    for _ in range(0, length):
         normal_cases.append(str(random.randint(INT_MIN, INT_MAX)))
```
Comment (revans2) [inline integration_tests/src/main/python/orc_cast_string_test.py:28]: Sorry, but NO. This test is not reproducible.  You are not setting a seed for random at all, which means you will get different data each time the test runs. So if there is a bug in here we will not be able to debug it if the test fails.

File: integration_tests/src/main/python/orc_cast_string_test.py [hunk 2]
```diff
@@ -56,7 +57,7 @@ def test_casting_string_to_float(spark_tmp_path, to_type):
         str(FLOAT_MAX), str(FLOAT_MIN), str(DOUBLE_MAX), str(DOUBLE_MIN)
     ]
     length = 2048
-    while len(normal_cases) < length:
+    for _ in range(0, length):
         normal_cases.append(str(random.uniform(DOUBLE_MIN, DOUBLE_MAX)))
```
Comment (revans2) [inline integration_tests/src/main/python/orc_cast_string_test.py:61]: Here too. At a minimum we need to set a seed

File: integration_tests/src/main/python/orc_cast_string_test.py [hunk 3]
```diff
@@ -76,3 +77,27 @@ def test_casting_string_to_float(spark_tmp_path, to_type):
...
+FIXME
+There is a strange case, if input string is "9808-02-30", intuitively, we should return null.
+However, CPU spark output 9808-02-39. Need to figure whether if it's a bug or a feature.
+'''
+def test_casting_string_to_date(spark_tmp_path):
+    orc_path = spark_tmp_path + '/orc_cast_string_to_date'
+    length = 2048
+    normal_cases = []
+    for _ in range(0, length):
+        year = str(random.randint(0, 9999)).zfill(4)
```
Comment (revans2) [inline integration_tests/src/main/python/orc_cast_string_test.py:92]: here too. We need a seed

**Review by revans2 — CHANGES_REQUESTED** [review]

**Comment by sinkinben** [comment]
Some issues when casting string to date (or timestamp).

**Corner cases**

```scala
// The string of date must be in a pattern of "YYYY-MM-DD".
//    + Four digits or more for the year. Years in the range 0000 to 9999 will be pre-padded by zero to ensure four digits. 
//    + Years outside that range will have a prefixed positive or negative symbol.
//    + Two digits for the month-of-year. This is pre-padded by zero to ensure two digits.
//    + Two digits for the day-of-month. This is pre-padded by zero to ensure two digits.

val data = Seq(
    ("", "2019-01-01"),  // 2019-01-01
    ("", "2011-01-00"),  // null
    ("", "2019-1-18"),   // null
    ("", "2019-01-3"),   // null
    ("", "19-01-01"),    // null
    ("", "19-1-1"),      // null
    ("", "2022-1-1"),    // null
    ("", "2020-13-01"),  // null
    ("", "2022-01-32"),  // pay attention to this, null
    ("", "9808-02-30"),  // pay attention to this, 9808-02-29
    ("", "2022-06-31"),  // pay attention to this, 2022-06-30
    ("", "2020-02-29"),
    ("", "0000-01-01"),
    ("", "0001-01-01")
)
val cols = Seq("str", "date_str")
val df=spark.createDataFrame(data).toDF(cols:_*).repartition(1)
df.printSchema()
df.show(100)
df.write.mode("overwrite").orc("/tmp/orc/data.orc")
```

And we cast `date_str` to `date`, the result of CPU spark is:
```shell
scala> var df = spark.read.schema("date_str date").orc("/tmp/orc/data.orc").show(100)
+----------+
|  date_str|
+----------+
|2019-01-01|
|      null|
|      null|
|      null|
|      null|
|      null|
|      null|
|      null|
|      null|
|9808-02-29|
|2022-06-30|
|2020-02-29|
|0000-01-01|
|0001-01-01|
+----------+
```

**Strange Issues**

There is a strange case, if input string is `"9808-02-30"`, intuitively, we should return null. However, CPU spark output `9808-02-29`. Need to figure whether if it's a bug or a feature.

And there exists the same problem when casting `string` to `timestamp`.

I have some discussions with member of Apache ORC, see https://github.com/apache/orc/issues/1237.

Schema evolution `spark.read.schema("date_str date")` is not the recommended method. In such case, spark recommends users to explicitly use `CAST` in SQL. That is:
```scala
var df = spark.read.schema("date_str string").orc("/tmp/orc/data.orc");
df.registerTempTable("table")
df.sqlContext.sql("select CAST(date_str as date) from table").show()
```

In some cases (e.g. `string -> date/timestamp`), keeping the same result of CPU is hard to implement on GPU. 

Shall we mark `string -> date/timestamp` unsupported in ORC reading? @revans2 @firestarman

**Comment by firestarman** [comment]
The casting about "9808-02-30" is not a bug, Spark supports year up to "9999" IIRC.

Which cases will GPU output different results than CPU ? @sinkinben

**Comment by sinkinben** [comment]
> The casting about "9808-02-30" is not a bug, Spark supports year up to "9999" IIRC.
> 
> Which cases will GPU output different results than CPU ? @sinkinben
> 
> I think we can support these castings, and with doc about these corner cases.

9808-02-30 is a invalid date, in cuDF we have two choice:
+ drop it and replce it with null by `ColumnVector.isTimestamp("%Y-%m-%d")`
+ keep it, calling `asTimestamp("%Y-%m-%d")` will convert it to 9808-03-01. Since there are 29 days in 9808-02, so the one more day will become 03-01.

**Comment by firestarman** [comment]
> 9808-02-30 is a invalid date, in cuDF we have two choice:
> 
> * drop it and replce it with null by `ColumnVector.isTimestamp("%Y-%m-%d")`
> * keep it, calling `asTimestamp("%Y-%m-%d")` will convert it to 9808-03-01. Since there are 29 days in 9808-02, so the one more day will become 03-01.

We can drop it and return null for this corner case, but remember to add doc for it.

**Pull request closed without merging**